File: source/downloads/code/LeetCode/Java/Surrounded-Regions.java
public class Solution {
    public void solve(char[][] board) {
        int i,j;
	int row = board.length;
	if(row==0) return;
	int col = board[0].length;
	for(i=0;i<row;i++){
	    // check first and last col
	    check(board,i,0,row,col);
	    check(board,i,col-1,row,col);
	}

	for(j=0;j<col;j++){
	    // check first and last row
	    check(board,0,j,row,col);
	    check(board,row-1,j,row,col);
	}

	// change all "O" to "x"
	for(i=0;i<row;i++)
	    for(j=0;j<col;j++)
		if(board[i][j]=='O') board[i][j] = 'X';

	// change "1" to "O"
	for(i=0;i<row;i++)
	    for(j=0;j<col;j++)
		if(board[i][j]=='1') board[i][j] = 'O';
	
    }

    public void check(char[][] board, int i, int j, int row, int col){

	if(board[i][j] == 'O'){
	    board[i][j] = '1';
	    if (i>1)	    
	        check(board,i-1,j,row,col);
	    if (j>1)
	        check(board,i,j-1,row,col);
	    if (i+1 < row)
	        check(board,i+1,j,row,col);
	    if (j+1 < col)
	        check(board,i,j+1,row,col);
	}
    }
}

File: downloads/code/LeetCode/Python/Surrounded-Regions.py
class Solution(object):
    def fill(self, board,queue,x, y):
        if x < 0 or x >= len(board) or y < 0 or y >= len(board[0]) or board[x][y] != 'O':
            return
        queue.append((x,y))
        board[x][y] = 'D'

    def bfs(self,board,queue,x,y):
        if board[x][y] == 'O':
            queue.append((x,y))
            self.fill(board,queue,x,y)

        # mark all neighbors
        while queue:
            cur = queue.pop(0)
            i = cur[0]
            j = cur[1]
            self.fill(board,queue,i-1,j)
            self.fill(board,queue,i,j-1)
            self.fill(board,queue,i,j+1)
            self.fill(board,queue,i+1,j)


    def solve(self, board):
        """
        :type board: List[List[str]]
        :rtype: void Do not return anything, modify board in-place instead.
        """

        if board == None or len(board) == 0:
            return

        row = len(board)
        col = len(board[0])
        queue = []

        for i in range(row):
            self.bfs(board,queue,i,0)
            self.bfs(board,queue,i,col-1)

        for i in range(col):
            self.bfs(board,queue,0,i)
            self.bfs(board,queue,row-1,i)

        for i in range(row):
            for j in range(col):
                if board[i][j] == 'O':
                    board[i][j] = 'X'
                elif board[i][j] == 'D':
                    board[i][j] = 'O'
        return
